File: backend/app/routes/copilot.py
"""AI Event Copilot — natural-language assistant that can act.

The LLM classifies the user's message into a structured action; the backend
executes it with the caller's role/tenant scope (create events, answer with real
stats/revenue, or draft marketing copy). Falls back to keyword parsing without a key.
"""
import json
import os
import re
from datetime import date, datetime
from typing import Optional, List

from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session
from sqlalchemy import func
from cerebras.cloud.sdk import Cerebras

from app.database import get_db
from app.core.deps import require_roles
from app.models import Event, Order, User, CopilotMessage

router = APIRouter(prefix="/api/copilot", tags=["Copilot"])
copilot_roles = require_roles("SUPER_ADMIN", "ORGANIZER", "SPONSOR")

api_key = os.getenv("CEREBRAS_API_KEY", "")
client = Cerebras(api_key=api_key) if api_key else None

SYSTEM = """You are EventoPro Copilot for an event-management platform.
Read the WHOLE conversation and gather details across multiple messages — the user
often provides the event title, client, date, venue, type and budget over several
turns. Remember everything already given; never ask again for something the user
already provided earlier in the conversation.

Reply ONLY with a compact JSON object, no markdown:
{"action": "<action>", "params": {...}, "reply": "<short friendly reply>"}

Valid actions:
- "create_event": params {title, client_name, event_date (YYYY-MM-DD), event_type, venue, budget}.
  As soon as you have title, client_name AND event_date (from anywhere in the
  conversation), use create_event with ALL known params. Only if one of those three
  is still missing, use action "clarify" and ask ONLY for the missing field(s).
- "get_stats": overview of the user's events.
- "get_revenue": ticket revenue so far.
- "draft_marketing": put the marketing copy in "reply".
- "answer": general help/answer in "reply".

Normalize dates to YYYY-MM-DD (e.g. "2026_08-01" -> "2026-08-01"). Do NOT invent
specific numbers in reply — the app appends real figures. Keep reply under 50 words.
Today is %س.""".replace("%س", date.today().isoformat())


SPONSOR_SYSTEM = """You are EventoPro Copilot for SPONSORS — you help brands decide
sponsorships. Read the WHOLE conversation and remember details already given.

Reply ONLY with a compact JSON object, no markdown:
{"action": "<action>", "params": {...}, "reply": "<short friendly reply>"}

Valid actions:
- "sponsor_roi": params {amount (number, rupees), event_title (optional)}. Use when the
  user asks what return / profit / leads / ROI they'd get for a given spend.
- "suggest_amount": params {event_title (optional)}. Use when the user asks how much
  they should sponsor.
- "answer": general help / advice in "reply".

Do NOT invent specific numbers in reply — the app appends the real projection.
Keep reply under 45 words. Today is %س.""".replace("%س", date.today().isoformat())


def _pick_event(db, title=None):
    q = db.query(Event).filter(Event.status.in_(["Upcoming", "In Progress"]))
    if title:
        e = q.filter(Event.title.ilike(f"%{title.strip()}%")).first()
        if e:
            return e
    return q.order_by(Event.expected_attendance.desc(), Event.event_date.asc()).first()


def _sponsor_roi(db, amount, title=None):
    amount = float(amount or 0)
    event = _pick_event(db, title)
    if not event:
        return {"kind": "roi", "error": "No upcoming events to project against yet."}
    attendance = int(event.expected_attendance or event.attendees_count or 500)
    budget = float(event.budget or 0)
    share = (amount / budget) if budget > 0 else 0.5
    rate = min(0.08, max(0.02, 0.02 + 0.06 * min(share, 1.0)))   # 2%–8% lead conversion
    impressions = attendance
    leads = max(0, round(attendance * rate))
    value_per_lead, value_per_impression = 2500, 25   # ₹ business value per lead / brand impression
    est_return = leads * value_per_lead + impressions * value_per_impression
    profit = est_return - amount
    roi_multiple = (est_return / amount) if amount > 0 else 0.0
    cost_per_lead = (amount / leads) if leads > 0 else 0.0
    return {
        "kind": "roi", "event": event.title, "amount": round(amount),
        "impressions": impressions, "leads": leads,
        "cost_per_lead": round(cost_per_lead), "estimated_return": round(est_return),
        "profit": round(profit), "roi_multiple": round(roi_multiple, 2),
    }


def _suggest_amount(db, title=None):
    event = _pick_event(db, title)
    if not event:
        return {"kind": "suggest", "error": "No upcoming events to recommend for yet."}
    attendance = int(event.expected_attendance or event.attendees_count or 500)
    rec = max(50000, round(attendance * 150 / 10000) * 10000)   # ~₹150 / expected attendee
    tiers = [
        {"name": "Bronze", "amount": round(rec * 0.5), "perks": "Logo on website + 2 passes"},
        {"name": "Silver", "amount": rec, "perks": "Exhibit booth + stage logo + 5 passes"},
        {"name": "Gold", "amount": rec * 2, "perks": "Headline branding + keynote mention + 10 passes"},
    ]
    return {"kind": "suggest", "event": event.title, "attendance": attendance,
            "recommended": rec, "tiers": tiers}


def _sponsor_fallback(message):
    m = message.lower()
    nums = re.findall(r"[\d,]+", m.replace("₹", ""))
    amount = None
    for n in nums:
        val = n.replace(",", "")
        if val.isdigit() and int(val) >= 1000:
            amount = int(val)
            break
    if amount:
        return {"action": "sponsor_roi", "params": {"amount": amount}, "reply": f"Here's the projected return on a ₹{amount:,} sponsorship:"}
    if "how much" in m or "suggest" in m or "recommend" in m or "should i" in m:
        return {"action": "suggest_amount", "params": {}, "reply": "Here's a recommended sponsorship for the top upcoming event:"}
    return {"action": "answer", "params": {},
            "reply": "Ask me things like ‘if I sponsor ₹5,00,000 how much profit?’ or ‘how much should I sponsor?’."}


class CopilotMsg(BaseModel):
    role: str
    content: str


class CopilotIn(BaseModel):
    message: str = Field(..., min_length=1)
    history: Optional[List[CopilotMsg]] = None


def _tenant_events_q(db, user):
    q = db.query(Event)
    if user.role != "SUPER_ADMIN" and user.tenant_id is not None:
        q = q.filter(Event.tenant_id == user.tenant_id)
    return q


def _do_stats(db, user):
    q = _tenant_events_q(db, user)
    total = q.count()
    upcoming = q.filter(Event.status == "Upcoming").count()
    completed = q.filter(Event.status == "Completed").count()
    budget = _tenant_events_q(db, user).with_entities(func.coalesce(func.sum(Event.budget), 0.0)).scalar() or 0
    return {"total_events": total, "upcoming": upcoming, "completed": completed, "total_budget": float(budget)}


def _do_revenue(db, user):
    q = db.query(func.coalesce(func.sum(Order.total_amount), 0.0)).filter(Order.status == "PAID")
    if user.role != "SUPER_ADMIN" and user.tenant_id is not None:
        q = q.join(Event, Event.id == Order.event_id).filter(Event.tenant_id == user.tenant_id)
    return {"ticket_revenue": float(q.scalar() or 0)}


def _parse_llm(text):
    text = text.strip()
    if text.startswith("```"):
        text = text.split("```", 2)[1] if "```" in text[3:] else text.strip("`")
        text = text.replace("json", "", 1).strip()
    try:
        return json.loads(text)
    except Exception:
        return {"action": "answer", "params": {}, "reply": text[:400]}


def _fallback(message):
    m = message.lower()
    if "revenue" in m or "sales" in m:
        return {"action": "get_revenue", "params": {}, "reply": "Here's your ticket revenue:"}
    if "stat" in m or "how many" in m or "overview" in m:
        return {"action": "get_stats", "params": {}, "reply": "Here's your event overview:"}
    return {"action": "answer", "params": {},
            "reply": "AI is not configured (set CEREBRAS_API_KEY). I can still show stats or revenue — try 'show my stats' or 'show revenue'."}


@router.post("")
def copilot(data: CopilotIn, current_user: User = Depends(copilot_roles), db: Session = Depends(get_db)):
    is_sponsor = current_user.role == "SPONSOR"
    system_prompt = SPONSOR_SYSTEM if is_sponsor else SYSTEM
    fallback = _sponsor_fallback if is_sponsor else _fallback

    # Load the user's stored conversation as context (source of truth for memory).
    stored = db.query(CopilotMessage).filter(
        CopilotMessage.user_email == current_user.email
    ).order_by(CopilotMessage.created_at).all()

    if client:
        try:
            messages = [{"role": "system", "content": system_prompt}]
            for h in stored[-12:]:
                messages.append({"role": ("assistant" if h.role == "assistant" else "user"),
                                 "content": (h.content or "")[:1000]})
            messages.append({"role": "user", "content": data.message})
            resp = client.chat.completions.create(model="gpt-oss-120b", messages=messages)
            plan = _parse_llm(resp.choices[0].message.content)
        except Exception:
            plan = fallback(data.message)
    else:
        plan = fallback(data.message)

    action = plan.get("action", "answer")
    params = plan.get("params", {}) or {}
    reply = plan.get("reply", "")
    result = None

    # ── Sponsor actions ──────────────────────────────────────────────────────
    if is_sponsor:
        if action == "sponsor_roi":
            result = _sponsor_roi(db, params.get("amount"), params.get("event_title"))
            if result.get("error"):
                reply = result["error"]
            elif not reply:
                reply = f"Projected return for a ₹{result['amount']:,} sponsorship of {result['event']}:"
        elif action == "suggest_amount":
            result = _suggest_amount(db, params.get("event_title"))
            if result.get("error"):
                reply = result["error"]
            elif not reply:
                reply = f"Recommended sponsorship for {result['event']}:"
        # persist + return below
        try:
            db.add(CopilotMessage(user_email=current_user.email, role="user", content=data.message[:2000]))
            db.add(CopilotMessage(user_email=current_user.email, role="assistant", content=(reply or "")[:2000]))
            db.commit()
        except Exception:
            db.rollback()
        return {"action": action, "reply": reply, "result": result}

    if action == "create_event":
        try:
            ev = Event(
                title=params.get("title"),
                client_name=params.get("client_name"),
                event_date=datetime.strptime(params["event_date"], "%Y-%m-%d").date(),
                event_type=params.get("event_type") or "Other",
                venue=params.get("venue"),
                budget=float(params.get("budget") or 0),
                status="Upcoming",
                tenant_id=(current_user.tenant_id if current_user.role != "SUPER_ADMIN" else None),
            )
            db.add(ev)
            db.commit()
            db.refresh(ev)
            result = {"event_id": ev.id, "title": ev.title}
            reply = reply or f"Created '{ev.title}' on {ev.event_date}."
        except Exception as e:
            action = "clarify"
            reply = "I need a title, client name, and date (YYYY-MM-DD) to create an event."
    elif action == "get_stats":
        result = _do_stats(db, current_user)
    elif action == "get_revenue":
        result = _do_revenue(db, current_user)

    # Persist the turn so the conversation continues next time (ChatGPT-style).
    try:
        db.add(CopilotMessage(user_email=current_user.email, role="user", content=data.message[:2000]))
        db.add(CopilotMessage(user_email=current_user.email, role="assistant", content=(reply or "")[:2000]))
        db.commit()
    except Exception:
        db.rollback()

    return {"action": action, "reply": reply, "result": result}


@router.get("/history")
def copilot_history(current_user: User = Depends(copilot_roles), db: Session = Depends(get_db)):
    """Return the user's saved Copilot conversation (oldest first)."""
    msgs = db.query(CopilotMessage).filter(
        CopilotMessage.user_email == current_user.email
    ).order_by(CopilotMessage.created_at).all()
    return [{"role": m.role, "content": m.content,
             "created_at": m.created_at.isoformat() if m.created_at else None} for m in msgs]


@router.delete("/history")
def clear_copilot_history(current_user: User = Depends(copilot_roles), db: Session = Depends(get_db)):
    """Start a new chat — clear the user's saved conversation."""
    db.query(CopilotMessage).filter(CopilotMessage.user_email == current_user.email).delete()
    db.commit()
    return {"status": "cleared"}

File: frontend/js/pages/copilot.js
registerPage('copilot', initCopilot);

// Conversation memory so the Copilot remembers details across messages.
let copilotHistory = [];

const COPILOT_SUGGESTIONS_MANAGER = [
    'Show my event stats',
    'How much ticket revenue so far?',
    'Create a Corporate event "Annual Meet" for client Acme on 2026-12-01, budget 500000',
    'Write a short marketing post for my next event',
];

const COPILOT_SUGGESTIONS_SPONSOR = [
    'If I sponsor ₹5,00,000, how much profit will I get?',
    'How much should I sponsor?',
    'What ROI can I expect for ₹2,00,000?',
    'Which event gives the best sponsorship value?',
];

function isSponsorCopilot() {
    return (window.currentUser || {}).role === 'SPONSOR';
}
function copilotSuggestions() {
    return isSponsorCopilot() ? COPILOT_SUGGESTIONS_SPONSOR : COPILOT_SUGGESTIONS_MANAGER;
}

async function initCopilot() {
    copilotHistory = [];
    const c = document.getElementById('page-container');
    c.innerHTML = `
        <div class="dashboard-header animate-fade-in stagger-1" style="margin-bottom:16px;display:flex;justify-content:space-between;align-items:flex-start;gap:12px;flex-wrap:wrap;">
            <div>
                <h3 style="font-size:1.6rem;color:var(--text-primary);">
                    <span class="material-icons-round" style="vertical-align:middle;color:var(--accent-primary);">auto_awesome</span>
                    AI Copilot
                </h3>
                <p style="color:var(--text-muted);">${isSponsorCopilot() ? 'Ask sponsorship questions — projected ROI, leads, and how much to invest.' : 'Ask in plain English — create events, check stats/revenue, draft marketing copy.'}</p>
            </div>
            <button class="btn btn-secondary btn-sm" onclick="newCopilotChat()"><span class="material-icons-round">add</span> New chat</button>
        </div>
        <div class="card card-glow fade-in stagger-2">
            <div class="card-body">
                <div id="copilot-log" style="min-height:280px;max-height:52vh;overflow-y:auto;display:flex;flex-direction:column;gap:12px;padding-bottom:8px;"></div>
                <div style="display:flex;flex-wrap:wrap;gap:8px;margin:12px 0;" id="copilot-suggestions">
                    ${copilotSuggestions().map(s => `<button class="chip" onclick="copilotAsk(this.textContent)" style="background:var(--bg-tertiary);border:1px solid var(--border-color);color:var(--text-secondary);padding:6px 12px;border-radius:20px;font-size:0.8rem;">${s}</button>`).join('')}
                </div>
                <form id="copilot-form" style="display:flex;gap:10px;">
                    <input id="copilot-input" class="form-input" placeholder="${isSponsorCopilot() ? 'e.g. if I sponsor ₹3,00,000 how much profit?' : 'e.g. create a wedding for client Sharma on 2026-11-20'}" autocomplete="off" style="flex:1;">
                    <button class="btn btn-primary" type="submit"><span class="material-icons-round">send</span></button>
                </form>
            </div>
        </div>`;
    document.getElementById('copilot-form').addEventListener('submit', (e) => {
        e.preventDefault();
        const v = document.getElementById('copilot-input').value.trim();
        if (v) copilotAsk(v);
    });

    // Load saved conversation (ChatGPT-style continuity).
    let past = [];
    try { past = await api.get('/copilot/history'); } catch (_) { past = []; }
    if (past.length) {
        past.forEach(m => {
            copilotBubble(m.role === 'assistant' ? 'assistant' : 'user', m.content);
            copilotHistory.push({ role: m.role, content: m.content });
        });
    } else if (isSponsorCopilot()) {
        copilotBubble('assistant', "Hi! I'm your Sponsorship Copilot. Ask me things like “if I sponsor ₹5,00,000, how much profit?” or “how much should I sponsor?” and I'll project reach, leads and ROI.");
    } else {
        copilotBubble('assistant', "Hi! I'm your Copilot. Try: “show my stats”, “how much revenue?”, or “create a Corporate event for client Acme on 2026-12-01”.");
    }
}

async function newCopilotChat() {
    try { await api.delete('/copilot/history'); } catch (_) { /* ignore */ }
    copilotHistory = [];
    const log = document.getElementById('copilot-log');
    if (log) log.innerHTML = '';
    copilotBubble('assistant', "New chat started. What would you like to do?");
    showToast('Started a new chat', 'info');
}

function copilotBubble(who, html) {
    const log = document.getElementById('copilot-log');
    if (!log) return;
    const mine = who === 'user';
    const div = document.createElement('div');
    div.style.cssText = `max-width:85%;padding:12px 14px;border-radius:14px;align-self:${mine ? 'flex-end' : 'flex-start'};` +
        (mine ? 'background:var(--accent-gradient);color:#fff;' : 'background:var(--bg-tertiary);color:var(--text-primary);border:1px solid var(--border-color);');
    div.innerHTML = html;
    log.appendChild(div);
    log.scrollTop = log.scrollHeight;
    return div;
}

function renderCopilotResult(result) {
    if (!result) return '';
    const inr = n => '₹' + Math.round(n || 0).toLocaleString('en-IN');
    // Sponsor: ROI projection
    if (result.kind === 'roi') {
        if (result.error) return '';
        const pos = result.profit >= 0;
        return `<div style="margin-top:10px;border-top:1px solid var(--border-color);padding-top:10px;font-size:0.85rem;">
            <div style="color:var(--text-muted);margin-bottom:6px;">Projection for <strong style="color:var(--text-primary);">${result.event}</strong> · spend ${inr(result.amount)}</div>
            <div style="display:grid;grid-template-columns:1fr 1fr;gap:8px;">
                <div>👁️ Impressions: <strong>${result.impressions.toLocaleString('en-IN')}</strong></div>
                <div>🎯 Est. leads: <strong>${result.leads.toLocaleString('en-IN')}</strong></div>
                <div>🧮 Cost / lead: <strong>${inr(result.cost_per_lead)}</strong></div>
                <div>💵 Est. return: <strong>${inr(result.estimated_return)}</strong></div>
                <div>📈 Est. profit: <strong style="color:${pos ? '#1A5FFF' : '#E4007C'};">${inr(result.profit)}</strong></div>
                <div>🚀 ROI: <strong style="color:${result.roi_multiple >= 1 ? '#1A5FFF' : '#FF2D95'};">${result.roi_multiple}×</strong></div>
            </div>
            <div style="color:var(--text-muted);font-size:0.72rem;margin-top:8px;">Illustrative estimate based on expected attendance & spend share.</div>
        </div>`;
    }
    // Sponsor: recommended amount + tiers
    if (result.kind === 'suggest') {
        if (result.error) return '';
        return `<div style="margin-top:10px;border-top:1px solid var(--border-color);padding-top:10px;font-size:0.85rem;">
            <div style="color:var(--text-muted);margin-bottom:8px;">For <strong style="color:var(--text-primary);">${result.event}</strong> (${(result.attendance || 0).toLocaleString('en-IN')} expected) — recommended <strong>${inr(result.recommended)}</strong>.</div>
            ${result.tiers.map(t => `<div style="display:flex;justify-content:space-between;gap:10px;padding:6px 0;border-bottom:1px solid var(--border-color);">
                <span><strong>${t.name}</strong> · <span style="color:var(--text-muted);">${t.perks}</span></span>
                <strong>${inr(t.amount)}</strong></div>`).join('')}
        </div>`;
    }
    if (result.total_events !== undefined) {
        return `<div style="margin-top:8px;font-size:0.85rem;color:var(--text-secondary);">
            📊 Events: <strong>${result.total_events}</strong> · Upcoming: <strong>${result.upcoming}</strong> · Completed: <strong>${result.completed}</strong> · Budget: <strong>₹${Math.round(result.total_budget).toLocaleString('en-IN')}</strong></div>`;
    }
    if (result.ticket_revenue !== undefined) {
        return `<div style="margin-top:8px;font-size:0.95rem;">💰 Ticket revenue: <strong>₹${Math.round(result.ticket_revenue).toLocaleString('en-IN')}</strong></div>`;
    }
    if (result.event_id) {
        return `<div style="margin-top:8px;font-size:0.85rem;color:#1A5FFF;">✅ Created event #${result.event_id}: <strong>${result.title}</strong></div>`;
    }
    return '';
}

async function copilotAsk(message) {
    const input = document.getElementById('copilot-input');
    if (input) input.value = '';
    copilotBubble('user', message);
    const thinking = copilotBubble('assistant', '<em style="color:var(--text-muted)">Thinking…</em>');
    try {
        const res = await api.post('/copilot', { message, history: copilotHistory.slice(-10) });
        thinking.innerHTML = (res.reply || '…') + renderCopilotResult(res.result);
        // Record the turn so the Copilot remembers context on the next message.
        copilotHistory.push({ role: 'user', content: message });
        copilotHistory.push({ role: 'assistant', content: res.reply || '' });
        if (copilotHistory.length > 20) copilotHistory = copilotHistory.slice(-20);
        // Refresh relevant page data if an event was created
        if (res.action === 'create_event' && res.result && res.result.event_id) {
            showToast('Event created by Copilot', 'success');
        }
    } catch (err) {
        thinking.innerHTML = `<span class="text-danger">${err.message || 'Copilot failed'}</span>`;
    }
    const log = document.getElementById('copilot-log');
    if (log) log.scrollTop = log.scrollHeight;
}
